### PR TITLE
runtime: cleanup and add config options

### PR DESCRIPTION
**Pull request by carllerche**
This patch finishes the cleanup as part of the transition to Tokio 0.2.
A number of changes were made to take advantage of having all Tokio
types in a single crate. Also, fixes using Tokio types from
`spawn_blocking`.

### Many threads, one resource driver

Previously, in the threaded scheduler, a resource driver (mio::Poll /
timer combo) was created per thread. This was more or less fine, except
it required balancing across the available drivers. When using a
resource driver from **outside** of the thread pool, balancing is
tricky. The change was original done to avoid having a dedicated driver
thread.

Now, instead of creating many resource drivers, a single resource driver
is used. Each scheduler thread will attempt to "lock" the resource
driver before parking on it. If the resource driver is already locked,
the thread uses a condition variable to park. Contention should remain
low as, under load, the scheduler avoids using the drivers.

### Add configuration options to enable I/O / time

New configuration options are added to `runtime::Builder` to allow
enabling I/O and time drivers on a runtime instance basis. This is
useful when wanting to create lightweight runtime instances to execute
compute only tasks.

### Bug fixes

The condition variable parker is updated to the same algorithm used in
`std`. This is motivated by some potential deadlock cases discovered by
`loom`.

The basic scheduler is fixed to fairly schedule tasks. `push_front` was
accidentally used instead of `push_back`.

I/O, time, and spawning now work from within `spawn_blocking` closures.

### Misc cleanup

The threaded scheduler is no longer generic over `P :Park`. Instead, it
is hard coded to a specific parker. Tests, including loom tests, are
updated to use `Runtime` directly. This provides greater coverage.

The `blocking` module is moved back into `runtime` as all usage is
within `runtime` itself.

### CODE REVIEW

**Comment by hawkw** [comment]
> This was more or less fine, except
> it required balancing across the available drivers. When using a
> resource driver from **outside** of the thread pool, balancing is
> tricky. The change was original done to avoid having a dedicated driver
> thread.

If memory serves, there were at least some cases where having many drivers resulted in a performance advantage for IO-bound programs? Has there been any assessment about the performance impact of this change?

File: tokio/src/runtime/park.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        // this call, we must perform a release operation that `park` can
+        // synchronize with. To do that we must write `NOTIFIED` even if `state`
+        // is already `NOTIFIED`. That is why this must be a swap rather than a
+        // compare-and-swap that returns if it reads `NOTIFIED` on failure.
+        match self.state.swap(NOTIFIED, SeqCst) {
+            EMPTY => return, // no one was waiting
+            NOTIFIED => return, // already unparked
+            PARKED_CONDVAR => self.unpark_condvar(),
+            PARKED_DRIVER => self.unpark_driver(),
+            _ => panic!("inconsistent state in unpark"),
```
Comment (hawkw) [inline tokio/src/runtime/park.rs]: nit: should this panic include the wrong state?

File: tokio/src/task/stack.rs [hunk 2]
```diff
@@ -28,7 +28,7 @@ impl<T: 'static> TransferStack<T> {
 
             // At this point, the queue_next field may also be used to track
             // whether or not the task must drop the join waker.
-            debug_assert_eq!(0, next & 1);
+            debug_assert_eq!(0, next & !1);
```
Comment (hawkw) [inline tokio/src/task/stack.rs]: was the previous assertion incorrect?

File: tokio/src/runtime/builder.rs [hunk 3]
```diff
@@ -82,6 +83,12 @@ impl Builder {
             // No task execution by default
             kind: Kind::Shell,
 
+            // I/O defaults to "off"
+            enable_io: false,
+
+            // Time defaults to "off"
+            enable_time: false,
+
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/builder.rs]: It only is when going via `Builder`, which **most** people should not have to.

File: tokio/src/task/stack.rs (hunk 2 above)
Comment (carllerche) [reply tokio/src/task/stack.rs]: yes, but debug assertions are not enabled when running loom in release mode...

File: tokio/src/runtime/blocking/mod.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+//! Abstracts out the APIs necessary to `Runtime` for integrating the blocking
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/mod.rs]: deadlines...

File: tokio/src/runtime/blocking/mod.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    mod task;
+
+    use crate::runtime::{self, Builder, io, time};
+
+    pub(crate) fn create_blocking_pool(
+        builder: &Builder,
+        spawner: &runtime::Spawner,
+        io: &io::Handle,
+        time: &time::Handle,
+        clock: &time::Clock,
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/mod.rs]: punting until we dig into combining all thread-locals into a single context object.

File: tokio/src/runtime/blocking/pool.rs [hunk 6]
```diff
@@ -214,71 +246,101 @@ impl Spawner {
...
-                    if shared.shutdown {
-                        // Work was produced, and we "took" it (by decrementing num_notify).
-                        // This means that num_idle was decremented once for our wakeup.
-                        // But, since we are exiting, we need to "undo" that, as we'll stay idle.
-                        shared.num_idle += 1;
-                        // NOTE: Technically we should also do num_notify++ and notify again,
-                        // but since we're shutting down anyway, that won't be necessary.
-                        break;
-                    }
+                inner.run();
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/pool.rs]: That is the main change. To move the logic into a fn to make it easier to wrap w/ the thread-local context.

File: tokio/src/runtime/blocking/pool.rs [hunk 7]
```diff
@@ -214,71 +246,101 @@ impl Spawner {
...
+            })
+            .unwrap();
+    }
+}
+
+impl Inner {
+    fn run(&self) {
+        let _io = io::set_default(&self.io_handle);
+
+        time::with_default(&self.time_handle, &self.clock, || {
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/pool.rs]: Punting for now.

**Pull request merged**